src/Sylius/Component/Rbac/Authorization/CachedPermissionMap.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Rbac\Authorization;
+
+use Doctrine\Common\Cache\Cache;
+use Sylius\Component\Rbac\Model\RoleInterface;
+
+/**
+ * Cached permission maps.
+ *
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class CachedPermissionMap implements PermissionMapInterface
+{
+    /**
+     * @var PermissionMapInterface
+     */
+    protected $map;
+
+    /**
+     * @var Cache
+     */
+    protected $cache;
+
+    /**
+     * @var int
+     */
+    protected $ttl;
+
+    /**
+     * @param PermissionMapInterface $map
+     * @param Cache $cache
+     * @param int $ttl
+     */
+    public function __construct(PermissionMapInterface $map, Cache $cache, $ttl = null)
+    {
+        $this->map = $map;
+        $this->cache = $cache;
+        $this->ttl = $ttl;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasPermission(RoleInterface $role, $permissionCode)
+    {
+        if ($this->cache->contains($this->getCacheKey($role))) {
+            return in_array($permissionCode, $this->cache->fetch($this->getCacheKey($role)));
```

Comment:
Little speed up would be to use `isset()` here, but it would require to little change in `foreach()` below.